cogs5e/models/sheet/coinpurse.py

```diff
@@ -67,6 +67,26 @@ def auto_convert(self, coins=None):
             raise InvalidArgument("You do not have enough coins to cover this transaction.")
         return coins
 
+    def auto_convert_up(self, coins=None):
```

Comment:
While these methods are logically similar, I think naming them `auto_convert_up` and `auto_convert_down` is a little bit of a foot-gun; semantically, these methods are doing pretty different things.

`auto_convert_down` is used to resolve a negative CoinsArgs delta to a valid amount of currency, while `auto_convert_up` probably shouldn't interact with CoinsArgs at all (I think this is the source of the confusion where you called this method with the bare CoinsArgs type above), and shuffle the currencies up in the `self` instance.

IMO, we should keep `auto_convert_down`'s name as is (or rename to something like `resolve_negative_delta`), and change this method to something like this: 
```py
    def consolidate_coins(self) -> CoinsArgs:  # no input CoinsArgs, returns the delta
        total_cp = self.total * 100
        new_pp = int(total_cp // 1000)
        total_cp -= coins.pp * 1000
        new_gp = int(total_cp // 100)
        total_cp -= coins.gp * 100
        new_ep = int(total_cp // 50)
        total_cp -= coins.ep * 50
        new_sp = int(total_cp // 10)
        total_cp -= coins.sp * 10
        new_cp = int(total_cp)

        delta = CoinsArgs(pp=new_pp - self.pp, gp=new_gp - self.gp, ep=new_ep - self.ep, sp=new_sp - self.sp, cp=new_cp - self.cp)
        self.set_currency(pp=new_pp, gp=new_gp, ep=new_ep, sp=new_sp, cp=new_cp)  # to ensure sync is fired

        return delta
```

The difference between this method being a primitive that mutates the data model, whereas I asked you to break `autoconvert_down` out, is that the user request for this primitive is a lot simpler. In natural conversation, it'd be something like this:

**autoconvert_down**:
- User: "I would like to remove 12 silver pieces" (`!g coins -12sp`)
- Bot: "you don't have enough silver, do you want to convert?" (resolve_currency, confirm)
- User: "yes"
- Bot: "ok, I can do this transaction by doing -2gp, +8sp instead" (autoconvert_down)
- User (implicitly): ok
- Bot: "ok, I applied the changes to your coinpurse" (update_currency, display delta)
- User: "cool, thanks"

**autoconvert_up**:
- User: "I would like to convert all my coins up" (`!g coins convert`)
- Bot: "ok, here you go" (autoconvert_up)
- Bot: "here's what changed:" (return and display delta)

... I don't know if that really makes sense, but my brain decided that it wanted to write a screenplay between a hypothetical user and a bot today, so here you go 😛

Reply:
Part of the issue I was having was the circular import for CoinsArgs when I tried to import it into coinpurse.py, thats why I had it sending a blank input CoinsArgs (I was going to ask you for a better solution today :p)

The rest of this makes a lot of sense, and I appreciate the screenplay!